tensorboard/backend/application.py

```diff
@@ -531,3 +562,58 @@ def _get_event_file_active_filter(flags):
   if inactive_secs < 0:
     return lambda timestamp: True
   return lambda timestamp: timestamp + inactive_secs >= time.time()
+
+
+class _DbModeMultiplexer(event_multiplexer.EventMultiplexer):
+  """Shim EventMultiplexer to use when in read-only DB mode.
+
+  In read-only DB mode, the EventMultiplexer is nonfunctional - there is no
+  logdir to reload, and the data is all exposed via SQL. This class represents
+  the do-nothing EventMultiplexer for that purpose, which serves only as a
+  conduit for DB-related parameters.
+
+  The load APIs raise exceptions if called, and the read APIs always
+  return empty results.
+  """
+  def __init__(self, db_uri, db_connection_provider):
+    """Constructor for `_DbModeMultiplexer`.
+
+    Args:
+      db_uri: A URI to the database file in use.
+      db_connection_provider: Provider function for creating a DB connection.
+    """
+    logger.info('_DbModeMultiplexer initializing for %s', db_uri)
+    super(_DbModeMultiplexer, self).__init__()
+    self.db_uri = db_uri
```

Comment:
~~Why does this serve as a conduit for DB-related parameters? IIUC nothing
uses this curently—the URI and connection provider are always pulled off
the context, not the multiplexer.~~

Oof, I see: there’s a `getattr`, which is why my `grep`s turned up
nothing. What do you think of either explicitly mentioning that these
are read in `TensorBoardWSGIApp`, or changing that logic above to be
something like

```python
db_uri = None
db_connection_provider = None
if isinstance(deprecated_multiplexer, _DbModeMultiplexer):
  db_uri = deprecated_multiplexer.db_uri
  db_connection_provider = deprecated_multiplexer.db_connection_provider
  deprecated_multiplexer = None  # don't pass no-op muxer into `TBContext`
```

to also resolve my gripe about including a non-functional multiplexer in
the context at all?

Reply:
See above response about the non-functional multiplexer.

I can change the getattr to two different `isinstance` checks (one per db multiplexer type) if you'd prefer that.  I was not trying to be terribly principled since this code is all going away soon, but I'm happy to make that change.

Reply:
Replaced getattrs with isinstance.